File: neetcode/blind75/graphs/clone_graph.py
# https://neetcode.io/problems/clone-graph

"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:

        if not node:
            return node

        # key: node from original graph, value: the copy
        created_nodes = dict()

        def copy(node):
            if node in created_nodes:
                return created_nodes[node]

            new_node = Node(node.val)

            created_nodes[node] = new_node
            for neighbor in node.neighbors:
                new_node.neighbors.append(copy(neighbor))

            return new_node
                
        return copy(node)

File: neetcode/neet150_JS_2/Graphs/clone-graph.js
// https://neetcode.io/problems/clone-graph

/*
create a Map for new node references from the original node:
    key: old node reference
    value: new node reference

recursive dfs
    if node === null:
        return null
    if Map has oldNode as a key
        return Map.get(oldNode) to return the reference to the new node

    create new Node from old node value

    iterate old node neighbors
        newNode.neighbors.push(this.rec(neigh, Map))

- Time: O(V + E)    // + E since must explore each neighbor for the node(vertice)
- Space: O(V)
*/

/**
 * // Definition for a Node.
 * class Node {
 *     constructor(val = 0, neighbors = []) {
 *       this.val = val;
 *       this.neighbors = neighbors;
 *     }
 * }
 */

class Solution {
    /**
     * @param {Node} node
     * @return {Node}
     */
    cloneGraph(node) {
        const createdMap = new Map()

        return this.rec(node, createdMap)
    }

    rec(node, map) {
        if (node === null) {
            return null
        }
        if (map.has(node)) {
            return map.get(node)
        }

        const newNode = new Node(node.val)
        map.set(node, newNode)

        for (let i = 0; i < node.neighbors.length; i ++) {
            newNode.neighbors.push(this.rec(node.neighbors[i], map))
        }

        return newNode
    }
}
